### PR TITLE
{compiler,lang}[GCCcore/12.3.0,foss/2023a] numba v0.58.1, LLVM v14.0.3 (patched for llvmlite)

### PR DESCRIPTION
**Pull request by schiotz**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by schiotz** [comment]
Note that this differs from previous versions of numba easyconfigs in the approach to llvmlite and LLVM.  Typically, llvmlite only works with a specific version of LLVM, which is typically not the one being used per default in the toolchain, and this has been handled by patching llvmlite to work with a different LLVM.  

Patching llvmlite to use LLVM 16 instead of LLVM 14 is certainly beyond me, and so is testing that it does not introduce weird errors. 

I chose the more pragmatic approach of linking llvmlite statically to LLVM 14.  Then LLVM can be demoted to a builddependency, and it no longer matters that the version is not consistent with other tools in the 2023a toolchain.  In my opinion, this approach is both easier and less prone to errors than the previous approach, at the price of a slightly larger module (99 MB versus 53 MB in my test).

Note also that numba 0.58.1 no longer install ``bin/pycc``, I see the same when pip-installing in a linux test machine.  The sanity check is updated accordingly.

**Comment by schiotz** [comment]
Test report by @schiotz
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
svol.fysik.dtu.dk - Linux CentOS Linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 3.6.8
See https://gist.github.com/schiotz/a13bee459eb3b1aa857e0815a08e80b5 for a full test report.

**Comment by Micket** [comment]
If we are going to make a special llvm for llvmlite we could just consider building the patched variant that llvmlite wants anyway.
https://llvmlite.readthedocs.io/en/latest/faqs.html#why-static-linking-to-llvm
they mention a set of patches, but we have always just ignored that until now:
https://github.com/numba/llvmlite/tree/main/conda-recipes

I had been thinking (and have mentioned in the past) that perhaps we should have made this a llvmlite+llvm bundle to avoid this problem, but i guess with these patches and having a separate module as a builddep + static linking would achieve the same.

File: easybuild/easyconfigs/l/LLVM/LLVM-14.0.3-GCCcore-12.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ specified code representation known as the LLVM intermediate representation
+ ("LLVM IR"). The LLVM Core libraries are well documented, and it is
+ particularly easy to invent your own language (or port an existing compiler)
+ to use LLVM as an optimizer and code generator."""
+
+toolchain = {'name': 'GCCcore', 'version': '12.3.0'}
+toolchainopts = {'cstd': 'gnu++11', 'pic': True}
+
+source_urls = ['https://github.com/llvm/llvm-project/releases/download/llvmorg-%(version)s/']
+sources = ['llvm-%(version)s.src.tar.xz']
```
Comment (Micket) [inline easybuild/easyconfigs/l/LLVM/LLVM-14.0.3-GCCcore-12.3.0.eb:17]: add source urls `https://github.com/numba/llvmlite/raw/v0.41.1/conda-recipes/` and the 3 patches;
```python3
patches = [
    'llvm14-clear-gotoffsetmap.patch',
    'llvm14-remove-use-of-clonefile.patch',
    'llvm14-svml.patch'
]
```

**Review by Micket — CHANGES_REQUESTED** [review]

**Comment by schiotz** [comment]
Thank you @Micket , I'll try this out.

I think the patches do not trigger static linking, and that should still be used to prevent conflicts with other packages using LLVM.  But good point that if there are some patches that should be used together with llvmlite, then we better apply them.

I'll look at this, but perhaps not today.

**Comment by schiotz** [comment]
Oh, the last patch is might be conda-specific, since conda builds stuff with MKL, and we do not do that.  It relates to the Intel SVML library (small vectors math library), which is used when building with Intel compilers, but probably not relevant (albeit perhaps not directly harmful) when building with foss.

Of course, since LLVM lives at GCCcore level, it should probably work with both GCC and Intel compilers, so perhaps the patch should still be applied ...

**Comment by Micket** [comment]
Sorry for not answering earlier, busy day. Turns out we can't do that trick with the patches and source urls, like i thought we could.

But anyway, if only clonefile remains, it seems to me that is juts a fix for apple platforms, which I think we can just ignore (as i think it's not well supported anyway). Sorry for not checking those patches more carefully.

**Comment by schiotz** [comment]
Your trick with the source URL actually worked, it is just the test suite that rejects it.

The SVML patch could be relevant if a version of numba is made for the Intel toolchain.  That is pt. stuck because the SciPy-bundle easyconfig has not been approved - and it may not be, as there are strange errors on older machines (#18875).

I'll look at bit more into the patches, too.

**Comment by schiotz** [comment]
I looked at the patches again.  

[llvm14-remove-use-of-clonefile.patch](https://github.com/numba/llvmlite/blob/main/conda-recipes/llvm14-remove-use-of-clonefile.patch) is clearly MacOS/Apple specific, and should probably be left out.

[llvm14-clear-gotoffsetmap.patch](https://github.com/numba/llvmlite/blob/main/conda-recipes/llvm14-clear-gotoffsetmap.patch) relates to the AArch64 architecture, this includes Apple Silicon but may not be limited to that, so I think it should be left in (although I do not think it is important).

[llvm14-svml.patch](https://github.com/numba/llvmlite/blob/main/conda-recipes/llvm14-svml.patch) seems to be important, as it fixes incorrect code generation when the Intel SVML library is used, probably when using the Intel toolchain.  Optimization using the SVML is enabled by default in the Intel compilers, and it is what gives the Intel compilers a significant edge for some codes (like my own Asap, where it gives more than a factor two in some cases).  In the worst case, including the patch does nothing (if no SVML code is emitted), but it may fix incorrect code generation.

For now, I'll include the two latter patches (and just add them to the EasyBuild repo, with a comment in the .eb file indicating the source.

**Comment by schiotz** [comment]
Test report by @schiotz
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
slid.fysik.dtu.dk - Linux CentOS Linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz, Python 3.6.8
See https://gist.github.com/schiotz/3a1946e3af3c8bafd6faf14b96ef4d76 for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
easybuild-c7.novalocal - Linux CentOS Linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/verdurin/24e7e4c3afece26673c91fc6f11aca7b for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ generoso
CORE_CNT=16

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node3123.skitty.os - Linux RHEL 8.8, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/boegel/80b45bd0b3024256c2a30222e1c00da2 for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ jsc-zen2
CORE_CNT=16

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Going in, thanks @schiotz!

**Pull request merged**